flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -972,24 +972,28 @@ private void releaseResources() {
 
         for (ResultPartitionWriter partitionWriter : consumableNotifyingPartitionWriters) {
             taskEventDispatcher.unregisterPartition(partitionWriter.getPartitionId());
-            if (isCanceledOrFailed()) {
-                partitionWriter.fail(getFailureCause());
-            }
         }
 
-        closeNetworkResources();
+        if (isCanceledOrFailed()) {
+            failAllResultPartitions();
+        }
+        closeAllResultPartitions();
```

Comment:
According to this code, we always should call `close` even after `fail` was called. But according to lamda which we pass to `TaskCanceler` we call `fail` without calling the `close`. Which way is more correct?

Reply:
In my understanding the fail method more likes release which mean the result partition can not be consumed and the close method closes all network resources (network buffers). We can consume a ResultPartition after the network buffers are recycled, for example, the blocking partition.